### PR TITLE
samples: psa_crypto: Create Certificate Signing Request

### PR DESCRIPTION
**Pull request by urutva**
This is a first step towards demonstrating provisioning of X.509
certificates. The device certificate signing request is created using
Mbed TLS X.509 APIs. The elliptic curve key `SECP256R1` used to sign the
CSR is generated inside TF-M and Mbed TLS is configured to use TF-M for
crypto services. This approach will ensure that private key never
leaves secure side.

The CSR is encoded in JSON and is printed on the console.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@linaro.org>

### CODE REVIEW

**Comment by microbuilder** [comment]
Good job getting this working and figuring out the issues with Zephyr + TF-M + MbedTLS 3.0!

If you can update the output so that we can easily copy/paste the CSR request, including proper line breaks, that would make this easier to use than the currernt output with `\n`:

```
{"CSR":"-----BEGIN CERTIFICATE REQUEST-----\nMIHrMIGQAgEAMC4xDzANBgNVBAoMBkxpbmFybzEbMBkGA1UEAwwSRGV2aWNlIENl\ncnRpZmljYXRlMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAENLcv1exBcbEE2b4c\n5933xMCx6WTLRR/jSpVSqHWyjE3xy0/CJiyQyQWy5Ewq6Z0R3zUbDobVnKEf/Prt\nIZq1KKAAMAwGCCqGSM49BAMCBQADSAAwRQIgaAlTPmrIaRO7myM2Qr+LNk9sagdO\njPGUqbz4oUWhUsICIQCuHADW6F2l4czv78BO5Nf+FHZEpjbI1+fA2aLzglOaiA==\n-----END CERTIFICATE REQUEST-----\n"}
qemu-system-arm: terminating on signal 2 from pid 31048 (<unknown process>)
```

**Comment by microbuilder** [comment]
You should be able to solve most of the stylistic issues with some variation of:

```bash
$ uncrustify -c $ZEPHYR_BASE/.uncrustify.cfg \
  -f samples/tfm_integration/psa_crypto/src/psa_crypto.c
```

That will output to stdout and you can replace the source file adding `-o filename.c`.

There is also an uncrustify plugin for VS Code that you can point to Zephyr's `.uncrustify.cfg` file and format files with that: https://marketplace.visualstudio.com/items?itemName=LaurentTreguier.uncrustify

File: samples/tfm_integration/psa_crypto/Kconfig [hunk 1]
```diff
@@ -19,6 +19,11 @@ config PSA_SHELL
 	help
 	  Enabling this option will make the 'psa' shell command available.
 
+config PSA_IMPORT_KEY
+	bool "Enable support for importing private key data"
+	help
+	  Enable support for importing private key data using PSA APIs.
```
Comment (microbuilder) [inline samples/tfm_integration/psa_crypto/Kconfig:25]: This description is probably a bit ambiguous, and would be clearer with
something like this to contrast with randomly generating the key.

```suggestion
	  Enable support for importing a pre-generated or randomly generated
	  private key using PSA APIs and PRIVATE_KEY_STATIC or
	  PRIVATE_KEY_RANDOM.
```

File: samples/tfm_integration/psa_crypto/Kconfig [hunk 2]
```diff
@@ -19,6 +19,11 @@ config PSA_SHELL
...
+
 choice
 	prompt "Private Key"
 	default PRIVATE_KEY_RANDOM
```
Comment (microbuilder) [inline samples/tfm_integration/psa_crypto/Kconfig:29]: Does this choice depend on `PSA_IMPORT_KEY` being set to true to be relevant? If so `depends on` should be added.

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c [hunk 3]
```diff
@@ -91,105 +107,421 @@ static psa_status_t crp_get_pub_key(psa_key_id_t key_id,
...
 	al_dump_log();
+	return status;
+}
 
-	/* Setup the key's attributes before the creation request. */
-	psa_set_key_id(&key_attributes, key_id);
-	psa_set_key_usage_flags(&key_attributes, key_usage);
-	psa_set_key_algorithm(&key_attributes, alg);
-	psa_set_key_type(&key_attributes, key_type);
+#else // CONFIG_PSA_IMPORT_KEY
```
Comment (microbuilder) [inline samples/tfm_integration/psa_crypto/src/psa_crypto.c:211]: Lines 211 and 316 have the same macro (`// CONFIG_PSA_IMPORT_KEY`), so it isn't clear which code block hands which case. Clarify which with a distinction between the two cases.

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c [hunk 4]
```diff
@@ -91,105 +107,421 @@ static psa_status_t crp_get_pub_key(psa_key_id_t key_id,
...
+    uint8_t priv_key_data[32] = { 0 };
+
+    LOG_INF("Generate rnadom data for private key");
+	al_dump_log();
+
+    psa_generate_random(priv_key_data, sizeof(priv_key_data));
+    LOG_INF("Random data generation for private key completed");
+	al_dump_log();
+
+#endif // CONFIG_PRIVATE_KEY_STATIC
```
Comment (microbuilder) [inline samples/tfm_integration/psa_crypto/src/psa_crypto.c:392]: Multiple usage of `CONFIG_PRiVATE_KEY_STATIC`

**Review by microbuilder — DISMISSED** [review]

File: samples/tfm_integration/psa_crypto/Kconfig (hunk 1 above)
Comment (urutva) [reply samples/tfm_integration/psa_crypto/Kconfig:25]: Fixed.

File: samples/tfm_integration/psa_crypto/Kconfig (hunk 2 above)
Comment (urutva) [reply samples/tfm_integration/psa_crypto/Kconfig:29]: Yes, fixed.

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c (hunk 3 above)
Comment (urutva) [reply samples/tfm_integration/psa_crypto/src/psa_crypto.c:211]: I thought adding the condition next to the `#else` would make it easier to understand. I've added a `!` before the condition in comment. Is that helpful?

The docstrings ([if](https://github.com/zephyrproject-rtos/zephyr/pull/40103/files/9393d7431d5928fc7d14eb5cc11e73408e943ab2#diff-eb71373f32415464bd1de047b52fafb9ae166ccf3e334e75ddd819c7bc1b7e9dR97) and [else](https://github.com/zephyrproject-rtos/zephyr/pull/40103/files/9393d7431d5928fc7d14eb5cc11e73408e943ab2#diff-eb71373f32415464bd1de047b52fafb9ae166ccf3e334e75ddd819c7bc1b7e9dR212)) for the functions define the functionality of the code block.

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c (hunk 4 above)
Comment (urutva) [reply samples/tfm_integration/psa_crypto/src/psa_crypto.c:392]: I've updated the comments next to the `if else` blocks. Hope it's easier to understand now.

**Review by microbuilder — APPROVED** [review]

**Review by theotherjimmy — APPROVED** [review]

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c [hunk 5]
```diff
@@ -191,6 +208,325 @@ static psa_status_t crp_gen_key_secp256r1(psa_key_id_t key_id,
...
+
+	psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+	psa_get_key_attributes(key_handle, &attributes);
+	mbedtls_x509write_csr_set_md_alg(&req, MBEDTLS_MD_SHA256);
+
+	LOG_INF("Adding subject name to CSR");
+	al_dump_log();
+
+	status = mbedtls_x509write_csr_set_subject_name(&req, "O=Linaro,CN=Device Certificate");
```
Comment (d3zd3z) [inline samples/tfm_integration/psa_crypto/src/psa_crypto.c:451]: This will need to be device-unique before real use, but is ok for now.

**Review by d3zd3z — APPROVED** [review]

File: samples/tfm_integration/psa_crypto/src/psa_crypto.c (hunk 5 above)
Comment (microbuilder) [reply samples/tfm_integration/psa_crypto/src/psa_crypto.c:451]: @urutva What we do with the test CA we've been using is to append a UUID here, which we could add to the sample.

For example, with `openssl` we do something like this to test:

```bash
$ openssl req -new -key USER.key -out USER.csr \
    -subj "/O=localhost/CN=$(uuidgen | tr '[:upper:]' '[:lower:]')"
```

Comment (microbuilder) [reply samples/tfm_integration/psa_crypto/src/psa_crypto.c:451]: We already generate random data from TF-M as part of this demo ... perhaps we could extract 128 bits from that stream, setting specific bits for a valid UUID as per RFC 4122, described here: https://www.cryptosys.net/pki/uuid-rfc4122.html

```
 The procedure to generate a version 4 UUID is as follows:

    Generate 16 random bytes (=128 bits)
    Adjust certain bits according to RFC 4122 section 4.4 as follows:
        set the four most significant bits of the 7th byte to 0100'B, so the high nibble is "4"
        set the two most significant bits of the 9th byte to 10'B, so the high nibble will be one of "8", "9", "A", or "B" (see Note 1).
    Encode the adjusted bytes as 32 hexadecimal digits
    Add four hyphen "-" characters to obtain blocks of 8, 4, 4, 4 and 12 hex digits
    Output the resulting 36-character string "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
```

**Pull request merged**